binary_search_tree/tree.py

```diff
@@ -6,50 +6,140 @@ def __init__(self, key, val = None):
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(log n)
     def add(self, key, value = None):
-        pass
+        if not self.root:
+            self.root = TreeNode(key, value)
+        else:
+            self.add_helper(self.root, key, value)
+       
 
-    # Time Complexity: 
-    # Space Complexity: 
+    def find_helper(self, current, key):
+        # Reached a leaf, nothing here
+        if current == None:
+            return None
+        # We found the value!
+        elif current.key == key:
+            return current.value
+        # Search to the left of the current node
+        elif key < current.key:
+            return self.find_helper(current.left, key)
+        # Search to the right of the current node
+        return self.find_helper(current.right, key)
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(log n)
     def find(self, key):
-        pass
+        return self.find_helper(self.root, key)
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
+    def inorder_helper(self, current, result):
+        if current:
+            # left 
+            self.inorder_helper(current.left, result)
+            # current
+            result.append({"key": current.key, "value": current.value})
+            # right
+            self.inorder_helper(current.right, result)
+
+        return result
 
-    # Time Complexity: 
-    # Space Complexity: 
     def inorder(self):
-        pass
+        tree = []
+        return self.inorder_helper(self.root, tree)
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)  
+    def preorder_helper(self, current, tree):
+        if current:
+            # current
+            tree.append({"key": current.key, "value": current.value})
+            # left
+            self.preorder_helper(current.left, tree)
+            # right
+            self.preorder_helper(current.right, tree)
+        return tree
 
-    # Time Complexity: 
-    # Space Complexity:     
     def preorder(self):
-        pass
+        tree = []
+        return self.preorder_helper(self.root, tree)
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n) 
+    def postorder_helper(self, current, tree):
+        if current:
+            #left
+            self.postorder_helper(current.left, tree)
+            #right
+            self.postorder_helper(current.right, tree)
+            #current
+            tree.append({"key": current.key, "value": current.value})
+
+        return tree
 
-    # Time Complexity: 
-    # Space Complexity:     
     def postorder(self):
-        pass
+        tree = []
+        return self.postorder_helper(self.root, tree)
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
+    def height_counter(self, current):
+        if not current:
+            return 0
+
+        return(max(self.height_counter(current.left), self.height_counter(current.right))) + 1
 
-    # Time Complexity: 
-    # Space Complexity:     
     def height(self):
```

Comment:
✨Nice! Time and space complexity would be O(n) if the tree is unbalanced! For a balanced tree, time and space complexity would be O(log n) because your recursive calls only go through half the nodes!